WIP: FT State Preparation Through Code Symmetries

inctechs opened the pull request:
## Description

We are trying to improve the fault-tolerant state preparation circuits by using code symmetries and/ or restructuring of circuits. 

## Checklist:

<!---
This checklist serves as a reminder of a couple of things that ensure your pull request will be merged swiftly.
-->

- [ ] The pull request only contains commits that are related to it.
- [ ] I have added appropriate tests and documentation.
- [ ] I have made sure that all CI jobs on GitHub pass.
- [ ] The pull request introduces no new warnings and follows the project's style guidelines.

inctechs committed:
create permutation file

inctechs committed:
add function to create steane state prep circuit

inctechs committed:
add function preparing Steane code CSS object

inctechs committed:
Correct wrong stabilizer generators

inctechs committed:
add option to check for logical 0

instead of just checking all flag qubits to be 0, one can also
check if the resulting measurement correspond to the logical
0 state

inctechs committed:
add set up for Simulation circuit

inctechs committed:
disable print statement removement

Print statements are automatically removed by the formatter.
Temporarily disable this for debugging purpose.

inctechs committed:
Merge remote-tracking branch 'origin/main' into ft-stateprep-qubit-perm

inctechs committed:
Merge remote-tracking branch 'origin/main' into ft-stateprep-qubit-perm

inctechs committed:
Merge remote-tracking branch 'origin/main' into ft-stateprep-qubit-perm

inctechs committed:
add utility function to get automorphism subgroup from generators

this utility function returns the whole automorphism subgroup of the
symmetric group based on the given generators of the automorphism
group

inctechs committed:
add function to evaluate fault set overlap based on permutations

This function as of now goes through all permutations and checks
how many faults overlap in the corresponding fault sets for each
permutation.

inctechs committed:
mandatory format changes (return -> Return)

inctechs committed:
Merge remote-tracking branch 'origin/main' into ft-stateprep-qubit-perm

inctechs committed:
Merge remote-tracking branch 'origin/main' into ft-stateprep-qubit-perm

pre-commit-ci[bot] committed:
🎨 pre-commit fixes

inctechs committed:
guide heuristic gauss based on penalty columns/ CNOTs

given a list of qubit indices, add a penalty to each corresponding col
of the cost matrix such that those CNOTs are applied later

inctechs committed:
update heuristic SPC function to allow guiding information

inctechs committed:
move overlap search out and turn it into a standalone function

pre-commit-ci[bot] committed:
🎨 pre-commit fixes

inctechs committed:
add method to permute qubits of StatePreparationCircuit

pehamTom committed:
Ignore missing library stubs sympy.

inctechs committed:
add method that checks overlap of fault sets to SPC class

inctechs committed:
add function to check mutual disjunctness

function takes two sets of state preparation circuits. one set of
current spcs and one of potential spcs. It then check which of the
potential new spcs has mutually disjoint fault set to all others.

inctechs committed:
add automorphism as optional property to CSSCode class

pehamTom committed:
Merge branch 'main' into ft-stateprep-qubit-perm

Signed-off-by: Tom Peham <tom.peham@tum.de>

pehamTom committed:
Simulation steane ftsp (#383)

## Description

This PR introduces a simulation class to simulate Steane-type
fault-tolerant state preparation circuits using four circuits.

Since this simulation method has a lot of overlapping functionality with
our previous simulation setup, this PR also refactors the common
functionality into the abstract class `NoisyNDFTStatePrepSimulator`. The
implementation of the old simulator is now in
`VerificationNDFTStatePrepSimulator`.

## Checklist:

<!---
This checklist serves as a reminder of a couple of things that ensure
your pull request will be merged swiftly.
-->

- [x] The pull request only contains commits that are related to it.
- [x] I have added appropriate tests and documentation.
- [ ] I have made sure that all CI jobs on GitHub pass.
- [x] The pull request introduces no new warnings and follows the
project's style guidelines.

---------

Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>

pehamTom committed:
Allow steane-type FTSP with only two circuits.

pehamTom committed:
Synthesis of state prep circuits in standard form.

pehamTom committed:
Checks in simulation dependent on state.

pehamTom committed:
Merge branch 'ft-stateprep-qubit-perm' into standard-form-commutations

inctechs committed:
remove debug prints

inctechs committed:
Merge branch 'ft-stateprep-qubit-perm' of https://github.com/cda-tum/mqt-qecc into ft-stateprep-qubit-perm

inctechs committed:
set up eval directory structure

inctechs committed:
add z_observable flag for z errors

pre-commit-ci[bot] committed:
🎨 pre-commit fixes

inctechs committed:
missing comma

inctechs committed:
Merge remote-tracking branch 'refs/remotes/origin/ft-stateprep-qubit-perm' into ft-stateprep-qubit-perm

inctechs committed:
Revert "add z_observable flag for z errors"

This reverts commit 705876b46bb8c1eab21a0c3e5d312325c868ba03.

pehamTom committed:
Canonical form circuits for Steane-type FTSP.

inctechs committed:
add binomial error calulation for SteaneNDFTStatePrepSimulator

inctechs committed:
prepare simulation

inctechs committed:
update cc_4_8_8 circuits

inctechs committed:
change cc_6_6_6 from qasm3 to qasm2

pehamTom committed:
Bugged implementation.

Broken for Golay Code.

inctechs committed:
add new headers for binomial errors in csv

inctechs committed:
remove falsly tracked files

pehamTom committed:
Second attempt.

pehamTom committed:
Fix bug in Steane Type measurement regarding measurement order.

pehamTom committed:
Merge remote-tracking branch 'origin/ft-stateprep-qubit-perm' into ft-stateprep-qubit-perm

pehamTom committed:
Random CNOT Permutation.

inctechs committed:
add circuits for cc_488 of distance 7

inctechs committed:
update sim file to include distance 7 cc488

inctechs committed:
Merge branch 'ft-stateprep-qubit-perm' of https://github.com/cda-tum/mqt-qecc into ft-stateprep-qubit-perm

inctechs committed:
correct crucial typo

pehamTom committed:
Speed up LUT creation.

pehamTom committed:
Merge remote-tracking branch 'origin/ft-stateprep-qubit-perm' into ft-stateprep-qubit-perm

pehamTom committed:
Remove old function.

pehamTom committed:
Remove another old function.

pehamTom committed:
Merge branch 'ft-stateprep-qubit-perm' into standard-form-commutations

pehamTom committed:
Fix index error in permutation generation.

pehamTom committed:
Steane-type check for Z errors.

pehamTom committed:
Remove pheno noise.

pehamTom committed:
Merge branch 'ft-stateprep-qubit-perm' into standard-form-commutations

pehamTom committed:
Noise on ancilla.

pehamTom committed:
Eval scripts for canonical Steane State Prep.

inctechs committed:
edit heuristic_gauss

add functionality that constructs a circuit based on a reference
circuit such that fault sets of both do not overlap

inctechs committed:
add function to find next error

building the circuit from behind we can track the current fault set
of the circuit live and hence decide which cnot can be build without
violating fault set overlaps.

inctechs committed:
change check mutual disjunctness function

before the function checked for overlaps in x as well as z fault
sets, but now one can choose to only check one (default x)
This was done as for bigger codes we cannot expect to have no
overlap in  both fault sets, so we have to consider fault sets
individually

inctechs committed:
change distance based fault set calculation

for the distance 7 case it is necessary to also consider
weight 3 errors of the single fault set, as otherwise we would allow
the situation where 2 weight 3 errors cancel but then a simple
weight 1 error can happen somewhere at the end of the circuit,
leaving a weight 4 error and hence violating fault tolerance

inctechs committed:
change build funciton for cnot circuit

make the function also work for circuits without hadamards

inctechs committed:
Merge remote-tracking branch 'origin/ft-stateprep-qubit-perm' into ft-stateprep-qubit-perm

inctechs committed:
reference based heuristic changed to reference fault sets and speed ups

Now we can construct a reference circuit based on fault sets that we
provide. Before only the fault set of the given reference SPC was used.
It is now possible to also find circuits based on other fault sets.

We have introduced some speed ups to the function. We save now errors
that have shown to be overlapping to prevent recalculation of those
errors. We also now only check the overlap with the new error and not
with the whole provisional fault set anymore.

inctechs committed:
define default for faulst set overlap check

if one of the two fault sets provided is empty, return False (no
overlap) by default.

pehamTom committed:
Fix canonical permutation.

pehamTom committed:
Merge remote-tracking branch 'origin/ft-stateprep-qubit-perm' into ft-stateprep-qubit-perm

pehamTom committed:
Removed old variable.

pehamTom committed:
Improve memory usage.

pehamTom committed:
Remove unnecessary stabilizer measurements in simulation.

pehamTom committed:
Indent

pehamTom committed:
Merge branch 'standard-form-commutations' into ft-stateprep-qubit-perm

pre-commit-ci[bot] committed:
🎨 pre-commit fixes

inctechs committed:
update octagon distance 7 simulation circuits

correct non-fault tolerant set of circuits

inctechs committed:
add scripts for distance 7 and private mac simulation

inctechs committed:
new circuit folder structure and add hexagonal d7 circuits

inctechs committed:
correct simulation script and expand it to include hexagonal d7 circuits

inctechs committed:
change C3 and C4 of hexd7 code to be reproduceable

inctechs committed:
heuristic indep of SPC (only code) add 2 error check

Remove dependency on SPC class and only work with CSSCode class

introduce possibility to also track 2 error events during the
construction to avoid that 2 error events of newly constructed circuit
cancel one error events of the reference circuit/ fault set

inctechs committed:
clean up StatePreparationCircuit Class add variables to heuristic

inctechs committed:
add variable descriptions

inctechs committed:
add fs overlap check to circuit and add Z logical to equivalence check

inctechs committed:
improve CNOT count of C4 for the distance 7 octagon code

inctechs committed:
remove unwanted csv files and change propabilities

inctechs committed:
add functionality to start z error sim

inctechs committed:
remove print statements when constructing heuristic circuit

pehamTom committed:
Fix bug in simulator circuit construction.

pehamTom committed:
Merge remote-tracking branch 'origin/main' into ft-stateprep-qubit-perm

inctechs committed:
reinstate gaussian elimination, copy advanced logic

return to state of original gaussian elimination and copy advanced logic
to a new function to work on it there.

inctechs committed:
clean up unused files and functions

inctechs committed:
set up GaussianElimination Class

implement a class for Gaussian elimination to then combine here
the basic elimination and the reference based elimination

inctechs committed:
Merge remote-tracking branch 'origin' into ft-stateprep-qubit-perm

inctechs committed:
refactor heuristic elimination

create GaussianElimination class that contains both the basic
elimination as well as the reference based elimination. This commit
contains several steps of refactoring.

inctechs committed:
add test file for gaussian elimination and first test

pre-commit-ci[bot] committed:
🎨 pre-commit fixes

inctechs committed:
remove Sympy from synthesis utils

inctechs committed:
Merge remote-tracking branch 'origin' into ft-stateprep-qubit-perm

inctechs committed:
Merge branch 'ft-stateprep-qubit-perm' of https://github.com/cda-tum/mqt-qecc into ft-stateprep-qubit-perm

inctechs committed:
solving mypy issues with sympy

inctechs committed:
solve issue with qsample and pytest

inctechs committed:
add integration test for basic elimination

inctechs committed:
fix typo

inctechs committed:
add more tests

inctechs committed:
add docstrings and typehints

inctechs committed:
fix first mypy issues

starting to fix mypy issues so that the CI runs through soon

inctechs committed:
add to API functions for both reference based and regular SP

pre-commit-ci[bot] committed:
🎨 pre-commit fixes

inctechs committed:
Merge remote-tracking branch 'origin' into ft-stateprep-qubit-perm

inctechs committed:
Merge branch 'ft-stateprep-qubit-perm' of https://github.com/cda-tum/mqt-qecc into ft-stateprep-qubit-perm

inctechs committed:
solve more mypy issues

inctechs committed:
fix final mypy issues in synthesis utils

inctechs committed:
remove old legacy function after refactoring

inctechs committed:
change name of penalty columns to penalty cnots

this is to avoid misunderstandings of what the variable represents

inctechs committed:
change name from penalty cols to penalty cnots on state_prep.py

inctechs committed:
remove mypy issues in synthesis_utils.py

inctechs committed:
remove legacy entry points to legacy function

replace the old legacy functions with the class api functions for the
basic elimination as well as the reference based elimination

inctechs committed:
remove legacy function form encoding.py and swap it with class structure

inctechs committed:
add pytest.ini to git ignore to keep locally neotest working

since parallel tests where added neotest struggles to read the json
reports from pytest as they are not necessarily well formatted with
parallel execution. However, locally I can run test also sequentially.

inctechs committed:
add required positional argument 'code' to GaussianElimination

pehamTom commented on src/mqt/qecc/circuit_synthesis/synthesis_utils.py:
```diff
@@ -90,73 +94,452 @@ def iterative_search_with_timeout(
+    return c_spcs
+
+
+def get_fs_based_on_d(spc: StatePrepCircuit) -> tuple[npt.NDArray[np.int8], npt.NDArray[np.int8]]:
```
I am realizing more and more that dealing with fault sets should be better supported since it is an essential primitive in a lot of our synthesis tasks. Do you think it makes sense to add more support for fault sets? Nothing for this PR but maybe we can open an issue and deal with that somewhat soon.

Regarding this method specifically, I would argue that the calling context knows the distance. The logic between the d=3,5,7 cases are different enough that you could have a dedicated method for each and avoid the branching here.

pehamTom commented on src/mqt/qecc/circuit_synthesis/synthesis_utils.py:
```diff
@@ -322,7 +705,8 @@ def build_css_circuit_from_cnot_list(n: int, cnots: list[tuple[int, int]], hadam
     """
     circ = QuantumCircuit(n)
-    circ.h(hadamards)
+    if hadamards:
```
Does qiskit not handle the case of empty arguments gracefully?

pehamTom commented on src/mqt/qecc/ecc_qiskit_wrapper.py:
This file has been removed from QECC. Please remove it.

pehamTom commented on test/python/circuit_synthesis/test_gaussian_elimination.py:
```diff
@@ -0,0 +1,274 @@
+from mqt.qecc.circuit_synthesis.synthesis_utils import CandidateAction, GaussianElimination
+from mqt.qecc.codes.css_code import CSSCode
+
+STEANE_CHECK_MATRIX = np.array([[1, 1, 0, 0, 1, 1, 0], [1, 0, 1, 0, 1, 0, 1], [0, 0, 0, 1, 1, 1, 1]], dtype=np.int8)
+STEANE_CODE = CSSCode(Hx=STEANE_CHECK_MATRIX)
```
Since these are only used in one fixture, I would define them in the fixture. 

Also you can get the code with `CSSCode.from_code_name("steane")`

pehamTom commented on test/python/circuit_synthesis/test_gaussian_elimination.py:
```diff
@@ -0,0 +1,274 @@
+    get_instance.failed_cnots = failed_cnots
+
+    # Act: Call the method
+    action = get_instance._get_candidate_action(i=0, j=1, costs=cost)
```
This applies to other parts of this file as well. Testing private methods is generally considered bad practice for a variety of reasons. I also think we don't need to test on this level of granularity, since we do coverage tests anyway that show whether the different scenarios or candidate actions are executed. 

You obviously have more experience with your implementation. Do you think it is possible to write test cases based on the "matrix" level that test the different scenarios?

pehamTom commented on test/python/circuit_synthesis/test_gaussian_elimination.py:
```diff
@@ -0,0 +1,274 @@
+def test_apply_cnot_to_matrix_updates_matrix_and_costs_correctly(get_instance):
+    """Verify the target column of the matrix is correctly updated after a CNOT."""
+    expected_matrix = np.array([[1, 1, 0, 0, 0, 1, 0], [1, 0, 1, 0, 0, 0, 1], [0, 0, 0, 1, 1, 1, 1]], dtype=np.int8)
+    get_instance._apply_cnot_to_matrix(i=0, j=4)
```
As above.

pehamTom commented on test/python/circuit_synthesis/test_gaussian_elimination.py:
```diff
@@ -0,0 +1,274 @@
+    used_qubit_index = 1
+    unused_qubit_index = 0
+    get_instance.used_columns.append(used_qubit_index)
+    masked_costs = get_instance._mask_out_used_qubits()
```
Same as above.

pehamTom reviewed (changes requested):
Thanks for reworking the logic of the CNOT synthesis part. I have left some comments throughout that hopefully help to improve the code. 

I think the part that requires the most refactoring is the tests. I have written a comment inline but to give more justification why I think it is not a good idea here to test private methods. Testing that the synthesis works correctly, should be independent of the class internals. We want to test whether the public interface of the class works correctly. I agree that this case is a bit difficult because there is a lot of logic hidden in the `GaussianElimination` class. I hope that we can find a way to formulate good synthesis cases without accessing the class internals. The matrices do not have to make sense from an error correction point of view anyway. Let me know if I can help with that.

Before I forget, the documentation also has to be done.

pehamTom committed:
Some minor refactoring.

pehamTom committed:
Removed unused method.

pehamTom committed:
Merge branch 'main' into ft-stateprep-qubit-perm

pehamTom committed:
Add ideal qubits to noise model.

pehamTom committed:
Added method to get unmeasured qubits of stim circuit.

pehamTom committed:
Start refactoring noisy simulation.

inctechs committed:
remove linter flag T201 for avoiding print removal

inctechs committed:
remove automorphism functionality from CSSCode class

inctechs committed:
remove sympy dependency from pyproject.toml

inctechs committed:
delete stateprep docs (previously removed on the master)

pehamTom committed:
Refactor Simulation functionality.

inctechs committed:
Merge branch 'ft-stateprep-qubit-perm' of https://github.com/cda-tum/mqt-qecc into ft-stateprep-qubit-perm

inctechs committed:
remove commented out code

inctechs commented on src/mqt/qecc/circuit_synthesis/synthesis_utils.py:
```diff
@@ -90,73 +94,452 @@ def iterative_search_with_timeout(
+        self.backtrack_required = False
+
+    def _validate_inputs(self) -> None:
+        """Here will be some checks that ensure that the reference based construction can actually be executed."""
```
Will be, weil ich nicht sicher war ob das hier alle checks sind oder ob man noch mehr machen sollte

inctechs commented on src/mqt/qecc/circuit_synthesis/synthesis_utils.py:
```diff
@@ -90,73 +94,452 @@ def iterative_search_with_timeout(
+            if self._handle_stagnation(costs_unused):
+                continue
+            candidate_pairs = self._get_candidate_pairs(costs_unused)
+            for i, j in candidate_pairs:
```
Yeah, so the idea of the candidate pair action enums was to make the control flow as clear as possible. Having the while-loop as the driving force of the algorithm and then the enums as different control flows for each iteration of the while loop, basically. So, I am open to making it clearer as this was also my objective, but I do not yet see how a different function would change that more?

inctechs committed:
rename EliminationCNOTSynthesizer and API functions

inctechs committed:
improve docstring of EliminationCNOTSynthesizer class

inctechs committed:
rename _handle_stagnation to _reset_if_stuck

inctechs commented on src/mqt/qecc/circuit_synthesis/synthesis_utils.py:
```diff
@@ -90,73 +94,452 @@ def iterative_search_with_timeout(
+        """
+        self._validate_inputs()
+        self._modify_matrix_structure()
+        self._ref_based_init()
```
and then initialize them here as class variables?, so inside _ref_based_init() set self.ref_x_fs, etc.?

inctechs committed:
update docstring of _is_reduced() method

inctechs commented on src/mqt/qecc/circuit_synthesis/synthesis_utils.py:
```diff
@@ -90,73 +94,452 @@ def iterative_search_with_timeout(
+        """Here will be some checks that ensure that the reference based construction can actually be executed."""
+        has_refs = bool(self.ref_x_fs.size or self.ref_z_fs.size)
+        has_code = self.code is not None
+        if has_refs ^ has_code:  # xor
```
learned something new and wanted to use it.
It is corrected now together with a logical error in this function

inctechs committed:
fix logical error in validate_inputs() method

inctechs committed:
fix type cast

inctechs committed:
remove _modify_matrix_structure() method

after internalt discussions we decided that it should be up to the user
to provide a different inital matrix to achieve a correct result. This
is because finding a solution does not depend on the matrix structure
but rather the speed of finding a solution.

inctechs committed:
separate greedy initialisation from fault set guided init

before the EliminationCNOTSynthesizer needed to be initialised with the
reference fault set, even though they were only used for one synthesis
method. Now the reference fault sets are passed directly to the
fault_set_guided_synthesis as arguments.

inctechs committed:
Merge branch 'main' into ft-stateprep-qubit-perm

inctechs committed:
Merge branch 'main' into ft-stateprep-qubit-perm

Signed-off-by: Erik Weilandt <48921674+inctechs@users.noreply.github.com>

pre-commit-ci[bot] committed:
🎨 pre-commit fixes

inctechs commented on src/mqt/qecc/circuit_synthesis/state_prep.py:
```diff
@@ -293,6 +345,346 @@ def gate_optimal_prep_circuit(
     return FaultyStatePrepCircuit(circ, code.x_distance // 2, code.z_distance // 2)
 
 
+def iter_cyclic_combinations(lists):
```
In this section, there are still some mypy issues. Should I go over this, or do you want to do so @pehamTom, as you can better judge which of the following is still needed and which is not in use anymore?

inctechs committed:
clean up repository and add personal playground directory

inctechs committed:
correct false initialization of fault set arrays

inctechs committed:
TEMPORARILY remove standard form features to pass CI

inctechs committed:
fix double initialisation due to wrong order

inctechs committed:
temp add unused code instance to pass test

long-term: reconsider if the EliminationCNOTSynthesizer class needs the
code instance. There might be situations where it is not necessary and
hence should be avoided to require the code at initialization.

inctechs committed:
adapt fault set handling to new PureFaultSet class structure

this might need bigger rethinking, so far I am handling the fault sets
as initially only as numpy arrays should that stay like this or change?

inctechs committed:
adapt method to new indexing of fault sets

there is no zero fault set anymore. the 0 fault set element is now the
single error fault sets which was before the '1' element in the list

inctechs committed:
Merge branch 'main' into ft-stateprep-qubit-perm

Signed-off-by: Erik Weilandt <48921674+inctechs@users.noreply.github.com>

inctechs committed:
Delete ecc_qiskit_wrapper.py

Deleted as requested by comment in code review.

inctechs committed:
fix mypy issues

inctechs committed:
add .qasm postfix to all circuit files to pass license hook

inctechs committed:
allow pickle despite security warnings

only self created files are used here

inctechs committed:
temporarily allow testing of private methods

this should be discussed again, if we want to keep the tests or only
test api functions here.
further changes: allow for scenario parameter as this help readability
and add module docstring.

inctechs committed:
Merge branch 'ft-stateprep-qubit-perm' of https://github.com/cda-tum/mqt-qecc into ft-stateprep-qubit-perm

inctechs committed:
fix error using variable prior to initialization

inctechs committed:
remove commented out code

inctechs committed:
fix edge case of wrong error type parsed

inctechs committed:
fix bug that propagation matrix does not get updated

inctechs committed:
add plus state simulation capability

inctechs committed:
Merge remote-tracking branch 'origin/main' into ft-stateprep-qubit-perm

inctechs committed:
prepare simulations

add circuits for rotated surface code d=5 and [[20,2,6]] code

inctechs committed:
change probabilities

inctechs committed:
add simulation results for 20,2,6 and rot surface code d5

inctechs committed:
Merge remote-tracking branch 'origin' into ft-stateprep-qubit-perm

inctechs committed:
change simulation propability

inctechs committed:
Merge branch 'main' of https://github.com/munich-quantum-toolkit/qecc into ft-stateprep-qubit-perm

inctechs committed:
add (repaired) stim circuit and check matrices of QECCs

Those are circuits from the Flag at Origin work that we compare against.

inctechs committed:
prepare simulation files for flag at origin sims

inctechs committed:
update mac bash script for 0.001 sims

inctechs committed:
add 0.001 results for square octagon distance 7 code

inctechs committed:
Merge remote-tracking branch 'origin/main' into ft-stateprep-qubit-perm

inctechs committed:
rename circuit and check matrix files for consitency

inctechs committed:
update fao simulation scripts to do remap of qubits

inctechs committed:
fix CSV files (comma was missing)

inctechs committed:
Merge branch 'main' of https://github.com/munich-quantum-toolkit/qecc into ft-stateprep-qubit-perm

pre-commit-ci[bot] committed:
🎨 pre-commit fixes

inctechs committed:
add draft for SteaneFTSP docs

inctechs committed:
WIP: save work, add current tests

inctechs committed:
add simulation files for linux cluster

inctechs committed:
update sim files to separate x and z sim

inctechs committed:
adjust p values

inctechs committed:
add fao updated simulations with resets

inctechs committed:
change simulation cap

now we can have a upper cap of --shots to avoid too long sims
